### PR TITLE
add some loss api

### PR DESCRIPTION
**Pull request by LokeZhou**
### PR Docs
https://github.com/PaddlePaddle/docs/pull/5928

### PR APIs

```bash
torch.nn.MSELoss
torch.nn.functional.binary_cross_entropy
torch.nn.BCELoss
torch.nn.L1Loss
torch.nn.Unfold
torch.nn.functional.unfold
```

### CODE REVIEW

File: paconvert/api_matcher.py [hunk 1]
```diff
@@ -3594,6 +3594,55 @@ def generate_code(self, kwargs):
         return code
 
 
+class MseLossMatcher(BaseMatcher):
+    def generate_code(self, kwargs):
```
Comment (zhwesky2010) [inline paconvert/api_matcher.py:3598]: 感觉这一块的重复度很高，是否可以统一成一个Matcher

**Comment by zhwesky2010** [comment]
> PaddlePaddle/docs#5928



> ### PR Docs
> [PaddlePaddle/docs#5928](https://github.com/PaddlePaddle/docs/pull/5928)
> 
> ### PR APIs
> ```shell
> torch.nn.MSELoss
> torch.nn.functional.binary_cross_entropy
> torch.nn.BCELoss
> torch.nn.L1Loss
> torch.nn.Unfold
> torch.nn.functional.unfold
> ```

这个文档的链接不对，不是loss的

File: paconvert/api_matcher.py (hunk 1 above)
Comment (LokeZhou) [reply paconvert/api_matcher.py]: done

**Comment by zhwesky2010** [comment]
PR描述里，docs直接放PR链接

File: paconvert/api_mapping.json [hunk 2]
```diff
@@ -8700,6 +8734,49 @@
...
+  "torch.nn.L1Loss": {
+    "Matcher": "SizeAverageMatcher",
+    "paddle_api": "paddle.nn.L1Loss",
+    "args_list": [
+      "size_average",
+      "reduce",
+      "reduction"
+    ]
+  },
+  "torch.nn.Unfold": {
```
Comment (zhwesky2010) [inline paconvert/api_mapping.json:8746]: 这个用genericmatcher有什么问题？

File: paconvert/api_mapping.json [hunk 3]
```diff
@@ -8700,6 +8734,49 @@
...
+      "stride"
+    ],
+    "kwargs_change": {
+      "kernel_size": "kernel_sizes",
+      "dilation": "dilations",
+      "padding": "paddings",
+      "stride": "strides"
+    }
+  },
+  "torch.nn.functional.unfold": {
```
Comment (zhwesky2010) [inline paconvert/api_mapping.json:8762]: 这个用genericmatcher有什么问题？

File: paconvert/api_mapping.json (hunk 2 above)
Comment (LokeZhou) [reply paconvert/api_mapping.json:8746]: kernel_size 参数 PyTorch 参数类型为 int、tuple(int) 或者 list(int)， Paddle 参数类型为 int 或者 list(int)，如果为tuple时，genericmatcher支持转写成list吗

Comment (LokeZhou) [reply paconvert/api_mapping.json:8746]: kernel_size 参数 pytorch支持tuple，paddle不支持，改为genericmatcher遇到tuple会报错

File: paconvert/api_mapping.json (hunk 3 above)
Comment (LokeZhou) [reply paconvert/api_mapping.json:8762]: 同torch.nn.Unfold

File: paconvert/api_matcher.py [hunk 4]
```diff
@@ -3359,6 +3359,30 @@ def generate_code(self, kwargs):
...
 
 
+class UnfoldMatcher(BaseMatcher):
+    def generate_code(self, kwargs):
+        if "kwargs_change" in self.api_mapping:
+            kwargs_change = self.api_mapping["kwargs_change"]
+            for key in list(kwargs_change.keys()):
+                if key in kwargs:
+                    if "input" not in key:
+                        if "(" in kwargs[key]:
```
Comment (zhwesky2010) [inline paconvert/api_matcher.py:3203]: 直接判断 if isinstance(kwargs[key] , ast.Tuple): 吧

File: paconvert/api_matcher.py [hunk 5]
```diff
@@ -3359,6 +3359,30 @@ def generate_code(self, kwargs):
...
+                            value = ast.literal_eval(kwargs[key])
+                            if isinstance(value, tuple):
+                                kwargs[key] = list(ast.literal_eval(kwargs[key]))
+                        else:
+                            kwargs[key] = "list({})".format(kwargs[key])
+                    kwargs[kwargs_change[key]] = kwargs[key]
+                    kwargs.pop(key)
+
+        if "paddings" not in kwargs:
```
Comment (zhwesky2010) [inline paconvert/api_matcher.py:3212]: 这个是默认值就不用单独设置了

File: paconvert/api_matcher.py [hunk 6]
```diff
@@ -3359,6 +3359,30 @@ def generate_code(self, kwargs):
         return GenericMatcher.generate_code(self, kwargs)
 
 
+class UnfoldMatcher(BaseMatcher):
+    def generate_code(self, kwargs):
+        if "kwargs_change" in self.api_mapping:
+            kwargs_change = self.api_mapping["kwargs_change"]
+            for key in list(kwargs_change.keys()):
+                if key in kwargs:
+                    if "input" not in key:
```
Comment (zhwesky2010) [inline paconvert/api_matcher.py:3202]: 这个由于不可能是tuple，也不用单独判断

File: paconvert/api_matcher.py [hunk 7]
```diff
@@ -3359,6 +3359,30 @@ def generate_code(self, kwargs):
         return GenericMatcher.generate_code(self, kwargs)
 
 
+class UnfoldMatcher(BaseMatcher):
```
Comment (zhwesky2010) [inline paconvert/api_matcher.py:3196]: 这个可以起一个通用的名字，这个主要功能是把tuple转成list：
可以叫Tuple2ListMatcher

File: paconvert/api_matcher.py [hunk 8]
```diff
@@ -3193,6 +3193,30 @@ def generate_code(self, kwargs):
...
+    def generate_code(self, kwargs):
```
Comment (zhwesky2010) [inline paconvert/api_matcher.py:3197]: 逻辑可以写成对每个kwargs遍历，判断是否kwargs，每个分支里再判断是否list，一共4个分支。用new_kwargs来接收kwargs，不然参数顺序会改变，导致代码风格不太好

```
for k in list(kwargs.keys()):
    if kwargs_change:
          if tuple:
          else:
    else:
          if tuple:
          else:
```

File: paconvert/api_matcher.py (hunk 5 above)
Comment (LokeZhou) [reply paconvert/api_matcher.py:3212]: 这个torch 默认是0；paddle默认是1

File: paconvert/api_matcher.py (hunk 8 above)
Comment (LokeZhou) [reply paconvert/api_matcher.py:3197]: done

File: paconvert/api_matcher.py (hunk 7 above)
Comment (LokeZhou) [reply paconvert/api_matcher.py]: done

File: paconvert/api_matcher.py (hunk 6 above)
Comment (LokeZhou) [reply paconvert/api_matcher.py]: done

File: paconvert/api_matcher.py (hunk 5 above)
Comment (LokeZhou) [reply paconvert/api_matcher.py]: done

File: paconvert/api_matcher.py (hunk 4 above)
Comment (LokeZhou) [reply paconvert/api_matcher.py]: 这个在generate_code直接这样判断似乎不起作用

File: paconvert/api_mapping.json [hunk 9]
```diff
@@ -9245,11 +9322,7 @@
     "paddle_api": "paddle.zeros_like",
     "args_list": [
       "input",
-      "dtype",
-      "layout",
-      "device",
-      "requires_grad",
-      "memory_format"
+      "index"
```
Comment (zhwesky2010) [inline paconvert/api_mapping.json:9325]: 这个改错了吧

File: paconvert/api_matcher.py [hunk 10]
```diff
@@ -3193,6 +3193,27 @@ def generate_code(self, kwargs):
...
 
 
+class Tuple2ListMatcher(BaseMatcher):
+    def generate_code(self, kwargs):
+        new_kwargs = {}
+        kwargs_change = self.api_mapping["kwargs_change"]
+        for k in list(kwargs.keys()):
+            if k in kwargs_change:
+                if "(" in kwargs[k] and isinstance(ast.literal_eval(kwargs[k]), tuple):
+                    new_kwargs[kwargs_change[k]] = list(ast.literal_eval(kwargs[k]))
```
Comment (zhwesky2010) [inline paconvert/api_matcher.py:3203]: 判断前面这个就可以，直接 'list({})'.format(kwargs[k]) 吧，尽量不要用eval执行这些逻辑

File: paconvert/api_matcher.py [hunk 11]
```diff
@@ -3193,6 +3193,27 @@ def generate_code(self, kwargs):
...
+                else:
+                    new_kwargs[kwargs_change[k]] = kwargs[k]
+            else:
+                if "(" in kwargs[k] and isinstance(ast.literal_eval(kwargs[k]), tuple):
+                    new_kwargs[k] = list(ast.literal_eval(kwargs[k]))
```
Comment (zhwesky2010) [inline paconvert/api_matcher.py:3208]: 同上

File: tests/test_nn_Unfold.py [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+
+        unfold = torch.nn.Unfold(kernel_size=(2, 3),stride=(2,2),dilation=(1,1),padding=(2,2))
+        result = unfold(input)
+
+        """
+    )
+    obj.run(pytorch_code, ["result"])
+
+
+def _test_case_7():
```
Comment (zhwesky2010) [inline tests/test_nn_Unfold.py:203]: 这个什么原因不支持，需要写明白

File: tests/test_nn_functional_unfold.py [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+              [0.0467307,  0.18104774, 0.9747845,  0.84306306]]]]
+              )
+
+        result = torch.nn.functional.unfold(input,kernel_size=(2, 3),stride=(2,2),dilation=(1,1),padding=(2,2))
+        """
+    )
+    obj.run(pytorch_code, ["result"])
+
+
+def _test_case_7():
```
Comment (zhwesky2010) [inline tests/test_nn_functional_unfold.py:195]: 这个什么原因不支持，需要写明白

Comment (LokeZhou) [reply tests/test_nn_functional_unfold.py:195]: done

File: tests/test_nn_Unfold.py (hunk 12 above)
Comment (LokeZhou) [reply tests/test_nn_Unfold.py:203]: done

File: paconvert/api_matcher.py (hunk 10 above)
Comment (LokeZhou) [reply paconvert/api_matcher.py]: 已更新判断逻辑

File: paconvert/api_matcher.py (hunk 11 above)
Comment (LokeZhou) [reply paconvert/api_matcher.py]: done

File: paconvert/api_mapping.json (hunk 9 above)
Comment (LokeZhou) [reply paconvert/api_mapping.json]: 已修正

**Review by zhwesky2010 — DISMISSED** [review]
LGTM

**Review by zhwesky2010 — APPROVED** [review]
